update-client-hub/modules/pal-psa/source/arm_uc_pal_psa_implementation.c

```diff
@@ -0,0 +1,548 @@
+    uint32_t                    header_hash_total;  // calculate on-the-fly over padded header,
+                                                    // body, and protected TLV
+    arm_uc_firmware_details_t   details;            // cached firmware details
+} firmware_update_ctx_t;
+
+/* non-secure firmware update context */
+static firmware_update_ctx_t firmware_update_ctx_nonsecure;
+                                            
+/**
+ * @brief Initialise the flash IAP API
+ *
+ * @param callback function pointer to the PAAL event handler
+ * @return Returns ERR_NONE on success.
+ *         Returns ERR_INVALID_PARAMETER on error.
+ */
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_Initialize(void (*callback)(uint32_t))
+{
+    arm_uc_error_t result = { .code = ERR_INVALID_PARAMETER };
+
+    int32_t active = arm_uc_flashiap_active_init();
+    if (active != ARM_UC_FLASHIAP_SUCCESS) {
+        UC_PAAL_ERR_MSG("arm_uc_flashiap_active_init() failed: %d\r\n", active);
+        result.code = ERR_UNSPECIFIED;
+        return result;
+    }
+
+    arm_uc_pal_psa_callback = callback;
+
+    /* signal done */
+    arm_uc_pal_psa_signal_internal(ARM_UC_PAAL_EVENT_INITIALIZE_DONE);
+
+    result.code = ERR_NONE;
+
+    return result;
+}
+
+/**
+ * @brief Get maximum number of supported storage locations.
+ *
+ * @return Number of storage locations.
+ */
+uint32_t ARM_UC_PAL_PSA_Mcuboot_GetMaxID(void)
+{
+    return MBED_CONF_UPDATE_CLIENT_STORAGE_LOCATIONS;
+}
+
+/**
+ * @brief Prepare the storage layer for a new firmware image.
+ * @details The storage location is set up to receive an image with
+ *          the details passed in the details struct.
+ *
+ * @param slot_id Storage location ID.
+ * @param details Pointer to a struct with firmware details.
+ * @param buffer Temporary buffer for formatting and storing metadata.
+ * @return Returns ERR_NONE on accept, and signals the event handler with
+ *         either DONE or ERROR when complete.
+ *         Returns ERR_INVALID_PARAMETER on reject, and no signal is sent.
+ */
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_Prepare(uint32_t slot_id,
+                                                   const arm_uc_firmware_details_t* details,
+                                                   arm_uc_buffer_t* buffer)
+{
+    arm_uc_error_t result = { .code = ERR_INVALID_PARAMETER };
+
+    /* validate input */
+    if (details &&
+        buffer && buffer->ptr &&
+        (slot_id < MBED_CONF_UPDATE_CLIENT_STORAGE_LOCATIONS)) {
+        UC_PAAL_TRACE("ARM_UC_PAL_PSA_Mcuboot_Prepare slot_id %" PRIu32 " details %p buffer %p",
+                      slot_id, details, buffer);
+        UC_PAAL_TRACE("FW size %" PRIu64, details->size);
+
+        /* TODO: Currently, only non-secure firmware update is supported. To support
+         *       secure/combined firmware update, it is necessary to read
+         *       header (and TLV) of active located in SPE, which is prohibited by
+         *       TF-M. */
```

Comment:
It shouldn't be necessary to calculate the hash in this file when using `psa_fwu_set_manifest` and `psa_fwu_query`. Depending on how the system is configured, it should be possible to do a full system update, updating secure and non-secure at the same time. In the new FOTA client it would also be possible to use component update to update each side individually.

Reply:
As above, TF-M's `psa_fwu_query()` doesn't return hash of active firmware. Besides, its `psa_fwu_set_manifest()` doesn't support. So I need to parse mcuboot and calculate hash of active firmware manually. But the solution can apply only on non-secure firmware and cannot on secure or combined one.

https://github.com/ARMmbed/mbed-os/blob/f6e165febec105c042137a4d1bf663a58968aac8/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/src/tfm_firmware_update_ipc_api.c#L183-L190